bam_markdup.c

```diff
@@ -467,6 +627,200 @@ static int add_duplicate(khash_t(duplicates) *d_hash, bam1_t *dupe) {
 }
 
 
+static inline int get_coordinate_positions(const char *qname, int *xpos, int *ypos) {
+    int sep = 0;
+    int pos = 0;
+
+    while (qname[pos]) {
+        if (qname[pos] == ':') {
+            sep++;
+
+            if (sep == 2) {
+                *xpos = pos + 1;
+            } else if (sep == 3) {
+                *ypos = pos + 1;
+            } else if (sep == 4) { // HiSeq style names
+                *xpos = *ypos;
+                *ypos = pos + 1;
+            }
+        }
+
+        pos++;
+    }
+
+    return sep;
+}
+
+/* Using the coordinates from the Illumina read name, see whether the duplicated read is
+   close enough (set by max_dist) to the original to be counted as optical.*/
+
+static int optical_duplicate(bam1_t *ori, bam1_t *dup, long max_dist) {
+    int ret = 0, seps;
+    char *original, *duplicate;
+    int oxpos = 0, oypos = 0, dxpos = 0, dypos = 0;
+
+
+    original  = bam_get_qname(ori);
+    duplicate = bam_get_qname(dup);
+
+    seps = get_coordinate_positions(original, &oxpos, &oypos);
+
+    if (!(seps == 3 || seps == 4)) {
+        fprintf(stderr, "[markdup] warning: cannot decipher read name %s for optical duplicate marking.\n", original);
```

Comment:
Might want to limit the number of these that get printed out, or it could get very spammy.

Reply:
I disagree.  It should either keep printing out error messages or fail completely.  I don't want it to not work silently.

I should probably mention the optical duplicates depends on the standard read names in the man page.

Reply:
That's a harsh punishment for someone who runs it with the wrong name format.  I presume you get one warning per alignment?

I'd have a static variable to count the number of warnings, and stop printing them after a while.

Reply:
If someone is running with the wrong name format and trying to get optical duplicates then it should probably not work at all.  No point in doing a full run if it is not going to work.

Alternatively, we could not throw an error while it is running and come up with a final total of errors at the end.  This could be useful for people using a mixture of data from different machines.  The downside is that is does not pinpoint what the problem read names are.

Reply:
I think I did something like that for the bwa check-qnames-match thing, but it must have been an early draft as I can't find it in the source code now. Something like

```c
for (all reads) {
    …
    if (strcmp(read_qname, mate_qname) != 0) {
        nerrors++;
        if (nerrors <= 10) fprintf(stderr, "names don't match %s != %s\n", read_qname, mate_qname);
    }
    …
}

if (nerrors > 10) fprintf(stderr, "And additional name mismatch errors: %llu more errors\n", nerrors - 10);
```

Reply:
I think something like that could work.  It gives examples of failures, a final report on numbers and doesn't produce huge amounts of spam.